### DISCUSSION TITLE
Extract the tangent operator and eigenvalue of spectral decomposition

### DISCUSSION THREAD

**Discussion by lmersel** [thelfer/tfel#318] [Q&A]
Hello,

I am a new user of MFront/MGIS and I am working on phase field modelling on FEniCSx. I am trying to reproduce Jeremy Bleyer's works on transferring data between MGIS and FEniCSx for some particular cases. 

For that, I reused a previous fracture behavior law called PhaseFieldSpectralSplit, implemented by Jeremy Bleyer's and you. I tried to extract the he tangent operator computed by mfront (mgis.behavior.K). But the array seems to be empty. I probably missed something to provide this results. With the .mfront file, I attached a small python test to handle with the different variables that I can extract.

At the same time, I would like to print the eigen values of the strain tensor. I managed to display the matrix containing the eigen vectors but the display of the vector containing the eigen values gives an error.  It is just to check in case of singular values.

Thank you for your help.

Best regards,

Lamia
[miehe-split.zip](https://github.com/thelfer/tfel/files/9626465/miehe-split.zip)

**Comment by thelfer** [comment]
Poke @bleyerj

**Reply by thelfer** [reply]
Hi @lmersel,

Thanks for your interest for `MFront` and `FEniCS`.

You explicitly request that the tangent operator shall not be computed here:

```python
it = mgis_bv.IntegrationType.IntegrationWithoutTangentOperator
```

Thus, the `K` array is not allocated.

Best,

Thomas

**Comment by lmersel** [comment]
Hi !
Indeed, I chose the wrong method, sorry for my mistake. I replaced with : 
`it = mgis_bv.IntegrationType.IntegrationWithConsistentTangentOperator` but the operator computed is zero at each iteration. I also tested with other `IntegrationType` (`PredictionWithTangentOperator`, `PredictionWithElasticOperator.`..) but it is still the same. Do I forget another command to compute it  correctly ?

Best,

Lamia

**Reply by thelfer** [reply]
Hi Lamia,
You shall print the values of `K` before updating your `MaterialDataManager`. Then it works :)
See the implementation of `update` here: https://github.com/thelfer/MFrontGenericInterfaceSupport/blob/master/src/MaterialDataManager.cxx
Best,
Thomas

**Reply by lmersel** [reply]
Indeed, it works perfectly ! Thank you !

**Reply by thelfer** [reply]
Good luck. Let us know if you make progresses

**Comment by lmersel** [comment]
Hello,

Currently, I am testing the behavior law on a bar in traction, in dynamic regime. I managed to switch the coupled variational formulation from internal fenicsx data to mfront data but the solution diverges. So I will apply the model on a quasi-static case already validated. It will be easier to investigate the issue. I will let you know the progress :)
